### PR TITLE
[FEATURE] Add support for highres images

### PR DESCRIPTION
**Pull request by wow-solution**
## Commit Message

By popular request, we are now adding high-resolution options to the existing responsive image rendering. The Bootstrap Package does not activate them by default, but it can be added relatively easily by configuration.

Example configuration to enable high-resolution support for the default variant for devices with a minimum device pixel ration from 1.5 and 2.

```
lib.contentElement.settings.responsiveimages.variants {
    default {
        sizes {
            1\.5x {
                multiplier = 1.5
            }
            2x {
                multiplier = 2
            }
        }
    }
}
```

The data for image rendering is now extended with a sub-set for high-resolution variants. If you did not overwrite the default templates, it will just work as soon as you add the configuration. If you have overwritten the default templates, it will still work as before but you need to add the new support for high-resolution images yourself. Please check the new updated templates and adjust your code if you want this support.

There is now a new key called `sizes` in every variant that should be rendered. This dataset consists of several entries each with a default key for a normal rendering (1x). This key will be added automatically and is always available. It is not possible to store sizes smaller than 1. Each size contains a multiplier, which is calculated with the base size.

Example: 
Base: 1000px

Minimum Device Pixel Ratio = 1.5
Multiplier = 1.25
Result: 1250px

Minimum Device Pixel Ratio = 2
Multiplier = 1.5
Result: 1500px

## Description
This changeset aims to make it possible to render high-resolution variants for responsive images. 
The patch is nonintrusive and is not changing existing behavior.

### Target Branches
- Master
- 10.x

### Prerequisites
* [x] Changes have been tested on TYPO3 v9.5 LTS
* [x] Changes have been tested on TYPO3 dev-master
* [x] Changes have been tested on PHP 7.2.x
* [x] Changes have been checked for CGL compliance `php-cs-fixer fix`

### Unit Tests Adapted
* [x] Unit Tests need to cover the new configurations
* [x] Add Unit-Tests for stacking configuration

**Changed applied to all responsive image renderings**
* [x] Background Images for Content Elements https://github.com/benjaminkott/bootstrap_package/blob/master/Resources/Private/Partials/ContentElements/Frame/General/BackgroundImage.html
* [x] Background Images for Carousel Items https://github.com/benjaminkott/bootstrap_package/blob/master/Resources/Private/Partials/ContentElements/Carousel/General/BackgroundImage.html
* [x] Simple Image Rendering https://github.com/benjaminkott/bootstrap_package/blob/master/Resources/Private/Partials/ContentElements/Media/Rendering/Image.html

## How to Test

The variant configuration for responsive images now accepts additional `sizes` configuration. You can add this additional configuration to your TypoScript setup. It will then render two additional images for the default variant.

### Example Configuration

```
lib.contentElement.settings.responsiveimages.variants.default {
    sizes {
        1\.5x {
            multiplier = 1.5
        }
        2x {
            multiplier = 2
        }
    }
}
```

### Expected Result

```HTML
<picture>
<source data-variant="default" data-maxwidth="530" media="(min-width: 1200px)" srcset="/fileadmin/_processed_/2/9/csm_typo3-book-backend-login_29be69984f.png 1x,/fileadmin/_processed_/2/9/csm_typo3-book-backend-login_6800abe34a.png 1.5x,/fileadmin/_processed_/2/9/csm_typo3-book-backend-login_e26ec99c45.png 2x">
<source data-variant="large" data-maxwidth="440" media="(min-width: 992px)" srcset="/fileadmin/_processed_/2/9/csm_typo3-book-backend-login_c25705d770.png 1x">
<source data-variant="medium" data-maxwidth="320" media="(min-width: 768px)" srcset="/fileadmin/_processed_/2/9/csm_typo3-book-backend-login_1e6b2adcf2.png 1x">
<source data-variant="small" data-maxwidth="500" media="(min-width: 576px)" srcset="/fileadmin/_processed_/2/9/csm_typo3-book-backend-login_9a0dfa6d8b.png 1x">
<source data-variant="extrasmall" data-maxwidth="374" srcset="/fileadmin/_processed_/2/9/csm_typo3-book-backend-login_dbad352291.png 1x">
<img src="/fileadmin/_processed_/2/9/csm_typo3-book-backend-login_28421744be.png" title="" alt="">
</picture>
```

```HTML
<style>
#carousel-backgroundimage-182-4 { background-image: url('/fileadmin/_processed_/4/6/csm_the-creative-exchange-373981-unsplash_896f6aa542.jpg'); } 
@media (min-width: 576px) { #carousel-backgroundimage-182-4 { background-image: url('/fileadmin/_processed_/4/6/csm_the-creative-exchange-373981-unsplash_856b1a5319.jpg'); } } 
@media (min-width: 768px) { #carousel-backgroundimage-182-4 { background-image: url('/fileadmin/_processed_/4/6/csm_the-creative-exchange-373981-unsplash_59eb9fec04.jpg'); } } 
@media (min-width: 992px) { #carousel-backgroundimage-182-4 { background-image: url('/fileadmin/_processed_/4/6/csm_the-creative-exchange-373981-unsplash_0f3525e7da.jpg'); } } 
@media (min-width: 1200px) { #carousel-backgroundimage-182-4 { background-image: url('/fileadmin/_processed_/4/6/csm_the-creative-exchange-373981-unsplash_8a9bdae6eb.jpg'); } } 
@media (-webkit-min-device-pixel-ratio: 1.5) and (min-width: 1200px), (min-resolution: 1.5dppx) and (min-width: 1200px) { #carousel-backgroundimage-182-4 { background-image: url('/fileadmin/_processed_/4/6/csm_the-creative-exchange-373981-unsplash_3d57bc4e86.jpg'); } } 
@media (-webkit-min-device-pixel-ratio: 2) and (min-width: 1200px), (min-resolution: 2dppx) and (min-width: 1200px) { #carousel-backgroundimage-182-4 { background-image: url('/fileadmin/_processed_/4/6/csm_the-creative-exchange-373981-unsplash_8dbc52b62b.jpg'); } } 
</style>
```

```HTML
<style>
#frame-backgroundimage-181 { background-image: url('/fileadmin/_processed_/e/e/csm_rawpixel-1046262-unsplash_b75e4ce0d1.jpg'); } 
@media (min-width: 576px) { #frame-backgroundimage-181 { background-image: url('/fileadmin/_processed_/e/e/csm_rawpixel-1046262-unsplash_a30250fc06.jpg'); } } 
@media (min-width: 768px) { #frame-backgroundimage-181 { background-image: url('/fileadmin/_processed_/e/e/csm_rawpixel-1046262-unsplash_b2da69437f.jpg'); } } 
@media (min-width: 992px) { #frame-backgroundimage-181 { background-image: url('/fileadmin/_processed_/e/e/csm_rawpixel-1046262-unsplash_0c14494133.jpg'); } } 
@media (min-width: 1200px) { #frame-backgroundimage-181 { background-image: url('/fileadmin/_processed_/e/e/csm_rawpixel-1046262-unsplash_9165aa58bf.jpg'); } } 
@media (-webkit-min-device-pixel-ratio: 1.5) and (min-width: 1200px), (min-resolution: 1.5dppx) and (min-width: 1200px) { #frame-backgroundimage-181 { background-image: url('/fileadmin/introduction/images/unsplash/rawpixel-1046262-unsplash.jpg'); } } 
@media (-webkit-min-device-pixel-ratio: 2) and (min-width: 1200px), (min-resolution: 2dppx) and (min-width: 1200px) { #frame-backgroundimage-181 { background-image: url('/fileadmin/introduction/images/unsplash/rawpixel-1046262-unsplash.jpg'); } } 
</style>
```

### CODE REVIEW

File: Classes/Utility/ImageVariantsUtility.php [hunk 1]
```diff
@@ -29,22 +30,27 @@ class ImageVariantsUtility
     protected static $defaultVariants = [
         'default' => [
             'breakpoint' => 1200,
-            'width' => 1100
+            'width' => 1100,
+            'widthRetina' => 2200
```
Comment (benjaminkott) [inline Classes/Utility/ImageVariantsUtility.php]: Please remove default variants, we only need to enable the user to configure retina images.
I would not like to deliver them by default.

File: Classes/Utility/ImageVariantsUtility.php [hunk 2]
```diff
@@ -29,22 +30,27 @@ class ImageVariantsUtility
...
         ],
         'large' => [
             'breakpoint' => 992,
-            'width' => 920
+            'width' => 920,
+            'widthRetina' => 1840
```
Comment (benjaminkott) [inline Classes/Utility/ImageVariantsUtility.php]: Please remove default variants, we only need to enable the user to configure retina images.
I would not like to deliver them by default.

File: Classes/Utility/ImageVariantsUtility.php [hunk 3]
```diff
@@ -29,22 +30,27 @@ class ImageVariantsUtility
...
         ],
         'medium' => [
             'breakpoint' => 768,
-            'width' => 680
+            'width' => 680,
+            'widthRetina' => 1360
```
Comment (benjaminkott) [inline Classes/Utility/ImageVariantsUtility.php]: Please remove default variants, we only need to enable the user to configure retina images.
I would not like to deliver them by default.

File: Classes/Utility/ImageVariantsUtility.php [hunk 4]
```diff
@@ -29,22 +30,27 @@ class ImageVariantsUtility
...
         ],
         'small' => [
             'breakpoint' => 576,
-            'width' => 500
+            'width' => 500,
+            'widthRetina' => 1000
```
Comment (benjaminkott) [inline Classes/Utility/ImageVariantsUtility.php]: Please remove default variants, we only need to enable the user to configure retina images.
I would not like to deliver them by default.

File: Classes/Utility/ImageVariantsUtility.php [hunk 5]
```diff
@@ -29,22 +30,27 @@ class ImageVariantsUtility
...
         ],
         'extrasmall' => [
-            'width' => 374
+            'width' => 374,
+            'widthRetina' => 748
```
Comment (benjaminkott) [inline Classes/Utility/ImageVariantsUtility.php]: Please remove default variants, we only need to enable the user to configure retina images.
I would not like to deliver them by default.

File: Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript [hunk 6]
```diff
@@ -168,22 +168,28 @@ lib.contentElement {
                 default {
                     breakpoint = 1200
                     width = 1100
+                    widthRetina = 2200
```
Comment (benjaminkott) [inline Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript]: As mentioned above, we should not configure it by default only enable the option.

File: Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript [hunk 7]
```diff
@@ -168,22 +168,28 @@ lib.contentElement {
...
                 }
+
```
Comment (benjaminkott) [inline Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript]: Remove empty line

File: Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript [hunk 8]
```diff
@@ -168,22 +168,28 @@ lib.contentElement {
                 default {
                     breakpoint = 1200
                     width = 1100
+                    widthRetina = 2200
                 }
+              
                 large {
                     breakpoint = 992
                     width = 920
+                    widthRetina = 1840
```
Comment (benjaminkott) [inline Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript]: As mentioned above, we should not configure it by default only enable the option.

File: Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript [hunk 9]
```diff
@@ -168,22 +168,28 @@ lib.contentElement {
...
                 }
                 medium {
                     breakpoint = 768
                     width = 680
+                    widthRetina = 1360
```
Comment (benjaminkott) [inline Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript]: As mentioned above, we should not configure it by default only enable the option.

File: Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript [hunk 10]
```diff
@@ -168,22 +168,28 @@ lib.contentElement {
...
                 }
                 small {
                     breakpoint = 576
                     width = 500
+                    widthRetina = 1000
```
Comment (benjaminkott) [inline Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript]: As mentioned above, we should not configure it by default only enable the option.

File: Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript [hunk 11]
```diff
@@ -168,22 +168,28 @@ lib.contentElement {
...
                 }
                 extrasmall {
                     breakpoint = unset
                     width = 374
+                    widthRetina = 748
```
Comment (benjaminkott) [inline Configuration/TypoScript/ContentElement/Helper/ContentElement.typoscript]: As mentioned above, we should not configure it by default only enable the option.

File: Resources/Private/Partials/ContentElements/Media/Rendering/Image.html [hunk 12]
```diff
@@ -3,10 +3,11 @@
...
         <f:variable name="width">{variant.width as integer}</f:variable>
+        <f:variable name="widthRetina">{variant.widthRetina as integer}</f:variable>
         <f:if condition="{width}">
-            <source data-variant="{name}" data-maxwidth="{width}" {f:if(condition: breakpoint, then: ' media="(min-width: {breakpoint}px)" ')}srcset="{f:uri.image(image: file, cropVariant: name, maxWidth: width)}">
+            <source data-variant="{name}" data-maxwidth="{width}" {f:if(condition: breakpoint, then: ' media="(min-width: {breakpoint}px)" ')}srcset="{f:uri.image(image: file, cropVariant: name, maxWidth: width)}{f:if(condition: widthRetina, then: ' , {f:uri.image(image: file, cropVariant: name, maxWidth: widthRetina)} 2x ')}">
         </f:if>
     </f:for>
     <img src="{f:uri.image(image: file, cropVariant: 'default', maxWidth: variants.default.width)}" title="{file.properties.title}" alt="{file.properties.alternative}">
 </picture>
-</html>
```
Comment (benjaminkott) [inline Resources/Private/Partials/ContentElements/Media/Rendering/Image.html]: Please ensure .editorconfig is loaded and an empty line is included at the end of each file.

File: Classes/Utility/ImageVariantsUtility.php [hunk 13]
```diff
@@ -20,7 +20,8 @@ class ImageVariantsUtility
      */
     protected static $allowedVariantProperties = [
         'breakpoint',
-        'width'
+        'width',
+        'widthRetina'
```
Comment (benjaminkott) [inline Classes/Utility/ImageVariantsUtility.php]: I am still unsure about this property, Do you think it really makes sense to set a with here?
What about just doing something like:

### Reducing configuration by just enabling different variations.

```
medium {
    breakpoint = 768
    width = 680
    highres = 1
    superhighres = 1
}
```

### Configuring variants with multiplier for more control about the image sizes

```
medium {
    breakpoint = 768
    width = 680
    variants {
       1x = 1
       2x = 1.5
       3x = 2
    }
}
```

**Comment by benjaminkott** [comment]
@wow-solution thx for the patch! It still needs some work. Please also check https://github.com/benjaminkott/bootstrap_package/blob/master/Tests/Unit/Utility/ImageVariantsUtilityTest.php to cover the new options.

File: Classes/Utility/ImageVariantsUtility.php (hunk 13 above)
Comment (wow-solution) [reply Classes/Utility/ImageVariantsUtility.php]: Great idea! I'll change my code and test this.

**Comment by benjaminkott** [comment]
@wow-solution I've adjusted the description and added notes about everything we need to change.

**Comment by wow-solution** [comment]
Just pushed my changes

**Comment by benjaminkott** [comment]
@wow-solution thx will check later

**Comment by wow-solution** [comment]
You used "ddpx" instead of "dppx" for backgroundimages resolution and "sizeDddpx" instead of "sizeDppx" for backgroundimages variable.

Also (already with dppx) the output always showes the image from @media screen and (min-width: 1200px) ...

**Pull request merged**